src/charm.py

```diff
@@ -258,6 +288,51 @@ def _update_auth_devices_keys(self) -> None:
                     auth_devices_keys
                 )
 
+    def _get_alert_rules_from_db(self, application: str):
+        database_url = (
+            self.external_url
+            + COS_REGISTRATION_SERVER_API_URL_BASE
+            + f"applications/{application}/alert_rules/"
+        )
+        try:
+            response = requests.get(database_url)
+            response.raise_for_status()
+            return response.json()
+        except requests.exceptions.RequestException as e:
+            logger.error(f"Failed to fetch {application} alert rules from '{database_url}': {e}")
+            return None
+
+    def _write_alert_rules_to_dir(self, path: str, alert_rules):
+        shutil.rmtree(path, ignore_errors=True)
+        mkdir(path)
+        for rules_file in alert_rules:
```

Comment:
When your iterable's name is plural, I suggest to use the singular for the loop variable: `alert_rule`.